k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java

```diff
@@ -342,103 +380,130 @@ private void updateRecipientExpanderVisibility() {
-        if (cachedCryptoStatus == null) {
-            ComposeCryptoStatusBuilder builder = new ComposeCryptoStatusBuilder()
-                    .setCryptoProviderState(cryptoProviderState)
-                    .setCryptoMode(currentCryptoMode)
-                    .setEnablePgpInline(cryptoEnablePgpInline)
-                    .setRecipients(getAllRecipients());
-
-            long accountCryptoKey = account.getCryptoKey();
-            if (accountCryptoKey != Account.NO_OPENPGP_KEY) {
-                // TODO split these into individual settings? maybe after key is bound to identity
-                builder.setSigningKeyId(accountCryptoKey);
-                builder.setSelfEncryptId(accountCryptoKey);
+        final ComposeCryptoStatus composeCryptoStatus = new ComposeCryptoStatusBuilder()
+                .setCryptoProviderState(cryptoProviderState)
+                .setCryptoMode(currentCryptoMode)
+                .setEnablePgpInline(cryptoEnablePgpInline)
+                .setRecipients(getAllRecipients())
+                .setSigningKeyId(accountCryptoKey)
+                .setSelfEncryptId(accountCryptoKey)
+                .build();
+
+        final String[] recipientAddresses = composeCryptoStatus.getRecipientAddresses();
+
+        new AsyncTask<Void,Void,RecipientAutocryptStatus>() {
+            @Override
+            protected RecipientAutocryptStatus doInBackground(Void... voids) {
+                if (cryptoProviderState != CryptoProviderState.OK) {
+                    return null;
+                }
+
+                return autocryptStatusInteractor.retrieveCryptoProviderRecipientStatus(
+                                getOpenPgpApi(), recipientAddresses);
             }
 
-            cachedCryptoStatus = builder.build();
+            @Override
+            protected void onPostExecute(RecipientAutocryptStatus recipientAutocryptStatus) {
+                if (recipientAutocryptStatus != null) {
+                    cachedCryptoStatus = composeCryptoStatus.withRecipientAutocryptStatus(recipientAutocryptStatus);
+                } else {
+                    cachedCryptoStatus = composeCryptoStatus;
+                }
+
+                redrawCachedCryptoStatusIcon();
+            }
+        }.execute();
+    }
+
+    private void redrawCachedCryptoStatusIcon() {
+        if (cachedCryptoStatus == null) {
+            throw new IllegalStateException("must have cached crypto status to redraw it!");
         }
 
+        recipientMvpView.setRecipientTokensShowCryptoEnabled(cachedCryptoStatus.isEncryptionEnabled());
+
+        CryptoStatusDisplayType cryptoStatusDisplayType = cachedCryptoStatus.getCryptoStatusDisplayType();
+        recipientMvpView.showCryptoStatus(cryptoStatusDisplayType);
+        recipientMvpView.showCryptoSpecialMode(cachedCryptoStatus.getCryptoSpecialModeDisplayType());
+    }
+
+    @Nullable
+    public ComposeCryptoStatus getCurrentCachedCryptoStatus() {
```

Comment:
What makes a crypto status "current"? Can there be a "current cached" crypto status? Maybe this could be just `getCachedCryptoStatus()`?

Reply:
This used to be `getCurrentCryptoStatus`, because it was constructed from the state of the recipient widgets. However, gathering this info now involves a roundtrip to the crypto provider, so the info we get is always only cached since we did that. It's the one that applies to the current recipients, but it is only cached in the sense that it *might* have changed at the crypto provider.

Reply:
`getLastFetchedCryptoStatus()` ?